File: Firmware/src/behaviors/condition.cpp
#include "condition.h"

namespace Behaviors
{
    /// <summary>
    /// Called by the Behavior Controller when a roll state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionIdle::checkTrigger(Modules::Accelerometer::RollState newState, int newFaceIndex) const {
        return newState == Modules::Accelerometer::RollState_OnFace || newState == Modules::Accelerometer::RollState_Crooked;
    }

    /// <summary>
    /// Called by the Behavior Controller when a roll state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionHandling::checkTrigger(Modules::Accelerometer::RollState newState, int newFaceIndex) const {
        return newState == Modules::Accelerometer::RollState_Handling;
    }

    /// <summary>
    /// Called by the Behavior Controller when a roll state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionRolling::checkTrigger(Modules::Accelerometer::RollState newState, int newFaceIndex) const {
        return newState == Modules::Accelerometer::RollState_Rolling;
    }

    /// <summary>
    /// Called by the Behavior Controller when a roll state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionCrooked::checkTrigger(Modules::Accelerometer::RollState newState, int newFaceIndex) const {
        return newState == Modules::Accelerometer::RollState_Crooked;
    }

    /// <summary>
    /// Called by the Behavior Controller when a roll state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionFaceCompare::checkTrigger(Modules::Accelerometer::RollState newState, int newFaceIndex) const {
        bool ret = false;
        if (newState == Modules::Accelerometer::RollState_OnFace) {
            if (flags & ConditionFaceCompare_Less) {
                // The flag says we should trigger if the face is less than our parameter
                ret = ret || (newFaceIndex < faceIndex);
            }
            if (flags & ConditionFaceCompare_Equal) {
                // The flag says we should trigger if the face is equal to our parameter
                ret = ret || (newFaceIndex == faceIndex);
            }
            if (flags & ConditionFaceCompare_Greater) {
                // The flag says we should trigger if the face is greater than our parameter
                ret = ret || (newFaceIndex > faceIndex);
            }
        }
        return ret;
    }

    /// <summary>
    /// Called by the Behavior Controller when a life state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionHelloGoodbye::checkTrigger(bool isHello) const {
        bool ret = false;
        if (flags & ConditionHelloGoodbye_Hello) {
            // The flag says we should trigger if the die is saying hello!
            ret = ret || isHello;
        }
        if (flags & ConditionHelloGoodbye_Goodbye) {
            // The flag says we should trigger if the die is saying goodbye!
            ret = ret || !isHello;
        }
        return ret;
    }

    /// <summary>
    /// Called by the Behavior Controller when a connection state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionConnectionState::checkTrigger(bool connected) const {
        bool ret = false;
        if (flags & ConditionConnectionState_Connected) {
            // The flag says we should trigger if the die just connected
            ret = ret || connected;
        }
        if (flags & ConditionConnectionState_Disconnected) {
            // The flag says we should trigger if the die just disconnected
            ret = ret || !connected;
        }
        return ret;
    }

    /// <summary>
    /// Called by the Behavior Controller when a battery state event happens to see if this condition should trigger
    /// </summary>
    bool ConditionBatteryState::checkTrigger(Modules::BatteryController::BatteryState newState) const {
        bool ret = false;
        if (flags & ConditionBatteryState_Ok) {
            // The flag says we should trigger is the battery is now Ok
            ret = ret || (newState == Modules::BatteryController::BatteryState_Ok);
        }
        if (flags & ConditionBatteryState_Low) {
            ret = ret || (newState == Modules::BatteryController::BatteryState_Low);
        }
        if (flags & ConditionBatteryState_Charging) {
            ret = ret || (newState == Modules::BatteryController::BatteryState_Charging);
        }
        if (flags & ConditionBatteryState_Done) {
            ret = ret || (newState == Modules::BatteryController::BatteryState_Done);
        }
        return ret;
    }

}

File: Central/Assets/Scripts/Behaviors/Condition.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using System.Runtime.InteropServices;

namespace Behaviors
{
    /// <summary>
    /// The types of conditions we support!
    /// </summary>
    public enum ConditionType : byte
    {
		[SkipEnumValue]
        Unknown = 0,
        [AdvancedEnumValue, Name("Die wakes up / sleeps")]
        HelloGoodbye,
        [Name("Die is picked up")]
        Handling,
        [Name("Die is rolling")]
		Rolling,
        [Name("Die roll is...")]
		FaceCompare,
        [Name("Die is crooked")]
		Crooked,
        [AdvancedEnumValue, Name("Bluetooth Event...")]
        ConnectionState,
        [AdvancedEnumValue, Name("Battery Event...")]
        BatteryState,
        [Name("Die is idle for...")]
        Idle,
    };

    /// <summary>
    /// The base struct for all conditions, stores a type identifier so we can tell the actual
    /// type of the condition and fetch the condition parameters correctly.
    /// </summary>
	public interface Condition
	{
		ConditionType type { get; set; }
    };

    /// <summary>
    /// Condition that triggers when the die is being handled
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionIdle
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.Idle;
        public byte padding1;
        public ushort repeatPeriodMs;
    };

    /// <summary>
    /// Condition that triggers when the die is being handled
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionHandling
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.Handling;
        public byte padding1;
        public byte padding2;
        public byte padding3;
    };

    /// <summary>
    /// Condition that triggers when the die is being rolled
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionRolling
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.Rolling;
        public byte padding1;
        public ushort repeatPeriodMs; // o means do NOT repeat
    };

    /// <summary>
    /// Condition that triggers when the die has landed by is crooked
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionCrooked
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.Crooked;
        public byte padding1;
        public byte padding2;
        public byte padding3;
    };

    /// <summary>
    /// Flags used to indicate how we treat the face, whether we want to trigger if the
    /// value is greater than the parameter, less, or equal, or any combination
    /// </summary>
    [System.Flags]
    public enum ConditionFaceCompare_Flags : byte
    {
        Less    = 1 << 0,
        Equal   = 1 << 1,
        Greater = 1 << 2
    };

    /// <summary>
    /// Condition that triggers when the die has landed on a face
    /// </summary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionFaceCompare
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.FaceCompare;
        public byte faceIndex;
        public ConditionFaceCompare_Flags flags;
        public byte paddingFlags;
    };

    /// <summary>
    /// Indicate whether the condition should trigger on Hello, Goodbye or both
    /// </summary>
    [System.Flags]
    public enum ConditionHelloGoodbye_Flags : byte
    {
        Hello      = 1 << 0,
        Goodbye    = 1 << 1
    };

    /// <summary>
    /// Condition that triggers on a life state event
    /// </sumary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionHelloGoodbye
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.HelloGoodbye;
        public ConditionHelloGoodbye_Flags flags;
        public byte padding1;
        public byte padding2;
    };

    /// <summary>
    /// Indicates when the condition should trigger, connected!, disconnected! or both
    /// </sumary>
    [System.Flags]
    public enum ConditionConnectionState_Flags : byte
    {
        Connected      = 1 << 0,
        Disconnected   = 1 << 1,
    };

    /// <summary>
    /// Condition that triggers on connection events
    /// </sumary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionConnectionState
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.ConnectionState;
        public ConditionConnectionState_Flags flags;
        public byte padding1;
        public byte padding2;
    };

    /// <summary>
    /// Indicates which battery event the condition should trigger on
    /// </sumary>
    [System.Flags]
    public enum ConditionBatteryState_Flags : byte
    {
        Ok        = 1 << 0,
        Low       = 1 << 1,
		Charging  = 1 << 2,
		Done      = 1 << 3
    };

    /// <summary>
    /// Condition that triggers on battery state events
    /// </sumary>
    [StructLayout(LayoutKind.Sequential, Pack = 1)]
    [System.Serializable]
    public class ConditionBatteryState
        : Condition
    {
        public ConditionType type { get; set; } = ConditionType.BatteryState;
        public ConditionBatteryState_Flags flags;
        public ushort repeatPeriodMs;
    };
}
